src/BenchmarkDotNet/Validators/BenchmarkClassValidator.cs

```diff
@@ -0,0 +1,30 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using BenchmarkDotNet.Extensions;
+using BenchmarkDotNet.Running;
+
+namespace BenchmarkDotNet.Validators
+{
+    public class BenchmarkClassValidator: IValidator
+    {
+        public static readonly IValidator Default = new BenchmarkClassValidator();
+        
+        public bool TreatsWarningsAsErrors => false;
+
+        private BenchmarkClassValidator() { }
+        
+        public IEnumerable<ValidationError> Validate(ValidationParameters validationParameters)
+        {
+            var notValid = validationParameters.Benchmarks.Where(x => x.Target.Type.IsGenericType)
+                                                          .Where(HasPrivateGenericArguments).ToList();
+            foreach (var benchmark in notValid)
+            {
+                yield return new ValidationError(true, $"Generic class {benchmark.Target.Type.GetDisplayName()} has non public generic argument(s)");
+            }
+        }
+
+        private bool HasPrivateGenericArguments(Benchmark benchmark) => benchmark.Target.Type.GetGenericArguments().Any(a => !a.IsPublic
+                                                                                                                          && !a.IsNestedPublic);
```

Comment:
could you please use `||` instead of `&&` in the condition?

Reply:
I guess i cant, cause public nested class has IsPublic=false and public non-nested has IsNested=false, so in case of OR condition we have (!a.IsPublic || !a.IsNestedPublic) = true for both public cases and for private case too - thats wrong.

Reply:
@Rizzen my bad, sorry! and thanks for the explanation! 

ps. this is a perfect case for unit test ;)

Reply:
No, I think i forgot about De Morgan's laws

Reply:
// Restore deleted comment
I guess i cant, cause public nested class has IsPublic=false and public non-nested has IsNestedPublic=false, so in case of OR condition we have (!a.IsPublic || !a.IsNestedPublic) = true for both public cases and for private case too -thats wrong

Reply:
@adamsitnik What do you think about two ways?

Reply:
@Rizzen I like `Any(a => !(a.IsPublic || a.IsNestedPublic))` but I really think that we should have unit tests for this. You should be able to write something similar to [these](https://github.com/dotnet/BenchmarkDotNet/blob/master/tests/BenchmarkDotNet.Tests/Validators/ExecutionValidatorTests.cs) tests for all 4 test cases (public/not public * nested/not nested)